pkg/diagnostics/pod/auth.go

```diff
@@ -0,0 +1,169 @@
+		return
+	}
+
+	// first try the secure connection in case they followed directions to secure the registry
+	// (https://docs.openshift.org/latest/install_config/install/docker_registry.html#securing-the-registry)
+	cacert, err := ioutil.ReadFile(d.MasterCaPath) // TODO: we assume same CA as master - better choice?
+	if err != nil {
+		r.Error("DP1008", err, fmt.Sprintf("Failed to read CA cert file %s:\n%v", d.MasterCaPath, err))
+		return
+	}
+	pool := x509.NewCertPool()
+	if !pool.AppendCertsFromPEM(cacert) {
+		r.Error("DP1009", err, fmt.Sprintf("Could not use cert from CA cert file %s:\n%v", d.MasterCaPath, err))
+		return
+	}
+	noSecClient := http.Client{
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+			return fmt.Errorf("No redirect expected")
+		},
+		Timeout: time.Second * 2,
+	}
+	secClient := noSecClient
+	secClient.Transport = &http.Transport{TLSClientConfig: &tls.Config{RootCAs: pool}}
+	if secError := processRegistryRequest(&secClient, fmt.Sprintf("https://%s:%s/v2/", registryHostname, registryPort), token, r); secError == nil {
+		return // made the request successfully enough to diagnose
+	} else if strings.Contains(secError.Error(), "tls: oversized record received") {
+		r.Debug("DP1015", "docker-registry not secured; falling back to cleartext connection")
+		if nosecError := processRegistryRequest(&noSecClient, fmt.Sprintf("http://%s:%s/v2/", registryHostname, registryPort), token, r); nosecError != nil {
+			r.Error("DP1013", nosecError, fmt.Sprintf("Unexpected error authenticating to the integrated registry:\n(%T) %[1]v", nosecError))
+		}
+	} else {
+		r.Error("DP1013", secError, fmt.Sprintf("Unexpected error authenticating to the integrated registry:\n(%T) %[1]v", secError))
+	}
+}
+
+// makes a request, handles some http/connection errors, returns any others
+func processRegistryRequest(client *http.Client, url string, token string, r types.DiagnosticResult) error {
+	req, _ := http.NewRequest("HEAD", url, nil)
+	req.SetBasicAuth("anyname", token)
+	if response, err := client.Do(req); err == nil {
+		switch response.StatusCode {
+		case 401, 403:
+			r.Error("DP1010", nil, "Service account token was not accepted by the integrated registry for authentication.\nThis indicates possible network problems or misconfiguration of the registry.")
+		case 200:
+			r.Info("DP1011", "Service account token was authenticated by the integrated registry.")
+		default:
+			r.Error("DP1012", nil, fmt.Sprintf("Unexpected status code from integrated registry authentication:\n%s", response.Status))
+		}
+		return nil
+	} else if strings.Contains(err.Error(), "net/http: request canceled") {
+		// (*url.Error) Head https://docker-registry.default.svc.cluster.local:5000/v2/: net/http: request canceled while waiting for connection
+		r.Error("DP1014", err, "Request to integrated registry timed out; this typically indicates network or SDN problems.")
+		return nil
+	} else {
```

Comment:
It's a small nit, but Golang convention is to avoid `else` like the plague and declare the var outside of the scope for these cases. This keeps indentation to a minimum. I'm not always a huge fan of it, but that's mostly what I try to do.